File: Sep/Day20/implement-router.cc
#include <bits/stdc++.h>
using namespace std;

struct Packet {
    int src, dst, ts;
    bool operator==(const Packet& other) const {
        return src == other.src && dst == other.dst && ts == other.ts;
    }
};

struct PacketHash {
    size_t operator()(const Packet& p) const {
        return ((p.src * 31 + p.dst) * 31 + p.ts);
    }
};

// ----------- Simple Version -------------
class RouterSimple {
    deque<Packet> dq;
    unordered_set<Packet, PacketHash> routerHashMap;
    int memoryLimit;

public:
    RouterSimple(int limit) : memoryLimit(limit) {}

    bool addPacket(int source, int destination, int timestamp) {
        Packet packet{source, destination, timestamp};
        if (routerHashMap.count(packet)) return false;

        if ((int)dq.size() == memoryLimit) forwardPacket();

        dq.push_back(packet);
        routerHashMap.insert(packet);
        return true;
    }

    vector<int> forwardPacket() {
        if (!dq.empty()) {
            Packet packet = dq.front();
            dq.pop_front();
            routerHashMap.erase(packet);
            return {packet.src, packet.dst, packet.ts};
        }
        return {};
    }

    int getCount(int destination, int startTime, int endTime) {
        int count = 0;
        for (auto& p : dq) {
            if (p.dst == destination && startTime <= p.ts && p.ts <= endTime)
                count++;
        }
        return count;
    }
};

// ----------- Optimized Version -------------
class RouterOptimized {
    deque<Packet> dq;
    unordered_set<Packet, PacketHash> routerHashMap;
    unordered_map<int, multiset<int>> destinationMap;
    int memoryLimit;

public:
    RouterOptimized(int limit) : memoryLimit(limit) {}

    bool addPacket(int source, int destination, int timestamp) {
        Packet packet{source, destination, timestamp};
        if (routerHashMap.count(packet)) return false;

        if ((int)dq.size() == memoryLimit) forwardPacket();

        dq.push_back(packet);
        routerHashMap.insert(packet);
        destinationMap[destination].insert(timestamp);
        return true;
    }

    vector<int> forwardPacket() {
        if (!dq.empty()) {
            Packet packet = dq.front();
            dq.pop_front();
            routerHashMap.erase(packet);

            destinationMap[packet.dst].erase(destinationMap[packet.dst].find(packet.ts));
            if (destinationMap[packet.dst].empty())
                destinationMap.erase(packet.dst);

            return {packet.src, packet.dst, packet.ts};
        }
        return {};
    }

    int getCount(int destination, int startTime, int endTime) {
        if (!destinationMap.count(destination)) return 0;

        auto& timestamps = destinationMap[destination];
        auto left = timestamps.lower_bound(startTime);
        auto right = timestamps.upper_bound(endTime);
        return distance(left, right);
    }
};

File: Sep/Day20/implement-router.js
class RouterSimple {
    constructor(memoryLimit) {
        this.dq = [];
        this.routerHashMap = new Set();
        this.memoryLimit = memoryLimit;
    }

    _packetKey(source, destination, timestamp) {
        return `${source},${destination},${timestamp}`;
    }

    addPacket(source, destination, timestamp) {
        let key = this._packetKey(source, destination, timestamp);
        if (this.routerHashMap.has(key)) return false;

        if (this.dq.length === this.memoryLimit) this.forwardPacket();

        this.dq.push([source, destination, timestamp]);
        this.routerHashMap.add(key);
        return true;
    }

    forwardPacket() {
        if (this.dq.length > 0) {
            let packet = this.dq.shift();
            let key = this._packetKey(...packet);
            this.routerHashMap.delete(key);
            return packet;
        }
        return [];
    }

    getCount(destination, startTime, endTime) {
        let count = 0;
        for (let [src, dst, ts] of this.dq) {
            if (dst === destination && startTime <= ts && ts <= endTime) count++;
        }
        return count;
    }
}

// ---------- Optimized ----------
class RouterOptimized {
    constructor(memoryLimit) {
        this.dq = [];
        this.routerHashMap = new Set();
        this.memoryLimit = memoryLimit;
        this.destinationMap = new Map(); // dest -> sorted array of timestamps
    }

    _packetKey(source, destination, timestamp) {
        return `${source},${destination},${timestamp}`;
    }

    addPacket(source, destination, timestamp) {
        let key = this._packetKey(source, destination, timestamp);
        if (this.routerHashMap.has(key)) return false;

        if (this.dq.length === this.memoryLimit) this.forwardPacket();

        this.dq.push([source, destination, timestamp]);
        this.routerHashMap.add(key);

        if (!this.destinationMap.has(destination)) {
            this.destinationMap.set(destination, []);
        }
        let arr = this.destinationMap.get(destination);
        let idx = this._bisectLeft(arr, timestamp);
        arr.splice(idx, 0, timestamp);
        return true;
    }

    forwardPacket() {
        if (this.dq.length > 0) {
            let packet = this.dq.shift();
            let key = this._packetKey(...packet);
            this.routerHashMap.delete(key);

            let [src, dst, ts] = packet;
            let arr = this.destinationMap.get(dst);
            let idx = this._bisectLeft(arr, ts);
            arr.splice(idx, 1);
            if (arr.length === 0) this.destinationMap.delete(dst);

            return packet;
        }
        return [];
    }

    getCount(destination, startTime, endTime) {
        if (!this.destinationMap.has(destination)) return 0;
        let arr = this.destinationMap.get(destination);
        let left = this._bisectLeft(arr, startTime);
        let right = this._bisectRight(arr, endTime);
        return right - left;
    }

    _bisectLeft(arr, x) {
        let lo = 0, hi = arr.length;
        while (lo < hi) {
            let mid = Math.floor((lo + hi) / 2);
            if (arr[mid] < x) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    }

    _bisectRight(arr, x) {
        let lo = 0, hi = arr.length;
        while (lo < hi) {
            let mid = Math.floor((lo + hi) / 2);
            if (arr[mid] <= x) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    }
}
